### PR TITLE
Support subsampling again after OSPRay 1.7 upgrade

### PR DESCRIPTION
**Pull request by tribal-tec**
This implies that the subsampled framebuffer which is used during interaction (and a
negative samples-per-pixel value), is now smaller than the window/viewport size.

Hence, clients are supposed to scale the image/framebuffer. This is done for the
braynsViewer using a texture.

### CODE REVIEW

**Comment by tribal-tec** [comment]
@rolandjitsu Needs upscaling of the JPEG image to the `viewport` size

File: apps/ui/BaseWindow.cpp [hunk 1]
```diff
@@ -300,8 +298,30 @@ void BaseWindow::display()
...
+            glVertex3f(0, 0, 0);
+            glTexCoord2f(0.f, 1.f);
+            glVertex3f(0, windowSize.y(), 0);
+            glTexCoord2f(1.f, 1.f);
+            glVertex3f(windowSize.x(), windowSize.y(), 0);
+            glTexCoord2f(1.f, 0.f);
+            glVertex3f(windowSize.x(), 0, 0);
+            glEnd();
+
+            glBindTexture(GL_TEXTURE_2D, 0);
```
Comment (hernando) [inline apps/ui/BaseWindow.cpp]: Maybe it's asking too much, but can you avoid using non GL3 core functions? You'll need shaders, but they are quite trivial.

File: apps/ui/BaseWindow.cpp [hunk 2]
```diff
@@ -406,6 +412,13 @@ void BaseWindow::create(const char* title, const size_t width,
         _screenSpaceProcessor.init(width, height);
 
     _registerKeyboardShortcuts();
+
+    glGenTextures(1, &_fbTexture);
+    glEnable(GL_TEXTURE_2D);
+    glBindTexture(GL_TEXTURE_2D, _fbTexture);
+    glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MIN_FILTER, GL_LINEAR);
+    glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MAG_FILTER, GL_LINEAR);
+    glBindTexture(GL_TEXTURE_2D, 0);
```
Comment (hernando) [inline apps/ui/BaseWindow.cpp]: No glDeleteTextures?

File: apps/ui/BaseWindow.cpp [hunk 3]
```diff
@@ -406,6 +412,13 @@ void BaseWindow::create(const char* title, const size_t width,
         _screenSpaceProcessor.init(width, height);
 
     _registerKeyboardShortcuts();
+
+    glGenTextures(1, &_fbTexture);
+    glEnable(GL_TEXTURE_2D);
+    glBindTexture(GL_TEXTURE_2D, _fbTexture);
+    glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MIN_FILTER, GL_LINEAR);
```
Comment (hernando) [inline apps/ui/BaseWindow.cpp]: MIN filter not needed?

File: engines/ospray/OSPRayFrameBuffer.cpp [hunk 4]
```diff
@@ -159,14 +165,42 @@ void OSPRayFrameBuffer::_unmapUnsafe()
...
+{
+    return _useSubsampling() ? _subsamplingFrameBuffer : _frameBuffer;
+}
+
+Vector2ui OSPRayFrameBuffer::_subsamplingSize() const
+{
+    return _frameSize / _subsamplingFactor;
+}
+
+OSPFrameBufferFormat OSPRayFrameBuffer::_ospFrameBufferFormat() const
```
Comment (hernando) [inline engines/ospray/OSPRayFrameBuffer.cpp]: Maybe free function with 1 parameter?

File: engines/ospray/OSPRayFrameBuffer.cpp [hunk 5]
```diff
@@ -90,42 +89,49 @@ void OSPRayFrameBuffer::_recreate()
...
+        ospCommit(_subsamplingFrameBuffer);
+    }
+}
+
 void OSPRayFrameBuffer::clear()
 {
     FrameBuffer::clear();
     size_t attributes = OSP_FB_COLOR | OSP_FB_DEPTH;
+    if (_subsamplingFrameBuffer)
+        ospFrameBufferClear(_subsamplingFrameBuffer, attributes);
```
Comment (hernando) [inline engines/ospray/OSPRayFrameBuffer.cpp]: Once created the _subsamplingFrameBuffer is never destroyed, is it?
Do you need to clear both buffers? Why not just _currentFB?

File: engines/ospray/OSPRayFrameBuffer.cpp [hunk 6]
```diff
@@ -90,42 +89,49 @@ void OSPRayFrameBuffer::_recreate()
...
     ospCommit(_frameBuffer);
+
+    _recreateSubsamplingBuffer();
+
     clear();
 }
 
+void OSPRayFrameBuffer::_recreateSubsamplingBuffer()
+{
+    ospRelease(_subsamplingFrameBuffer);
```
Comment (hernando) [inline engines/ospray/OSPRayFrameBuffer.cpp]: Just for confirmation, ospRelease(nullptr) is safe, isn't it?

File: engines/ospray/OSPRayFrameBuffer.cpp [hunk 7]
```diff
@@ -54,6 +51,8 @@ void OSPRayFrameBuffer::enableDeflectPixelOp()
     {
         ospCommit(_pixelOp);
         ospSetPixelOp(_frameBuffer, _pixelOp);
+        if (_subsamplingFrameBuffer)
+            ospSetPixelOp(_subsamplingFrameBuffer, _pixelOp);
```
Comment (hernando) [inline engines/ospray/OSPRayFrameBuffer.cpp]: I don't know the logic of the pixelop, but it seems to me that Deflect may not work properly after subsampling has been enabled.

**Comment by karjonas** [comment]
Is it feasible to return a framebuffer of the correct size so that the client does not need to do any scaling?

**Comment by hernando** [comment]
I guess that's doable using an ispc kernel. However for network streaming I think it's preferable to ask the client to do the upscaling (which can also be hardware accelerated in the display side).

Even without network communication, I would bet that doing the upscaling during display brings better performance as any GPU has better memory bandwidth for such an operation than a CPU and you still have to upload the pixels to the GPU, so the smaller the buffer, the better.

File: apps/ui/BaseWindow.cpp (hunk 1 above)
Comment (tribal-tec) [reply apps/ui/BaseWindow.cpp]: does this really help? I bluntly took what the ospray example viewer is doing, so no efforts spend so far. GL is dead anyways, so using legacy GL is not too much of a difference to GL3.

File: apps/ui/BaseWindow.cpp (hunk 2 above)
Comment (tribal-tec) [reply apps/ui/BaseWindow.cpp]: that's probably missing, thanks

File: engines/ospray/OSPRayFrameBuffer.cpp (hunk 5 above)
Comment (tribal-tec) [reply engines/ospray/OSPRayFrameBuffer.cpp]: `clear()` shall clear everything, also the accumulation buffer in the 'normal' framebuffer. `_currentFB()` at the time when `clear()` is called will only clear the subsampling buffer.

File: engines/ospray/OSPRayFrameBuffer.cpp (hunk 6 above)
Comment (tribal-tec) [reply engines/ospray/OSPRayFrameBuffer.cpp]: yes, it is. hence I removed some of the checks I found.

File: engines/ospray/OSPRayFrameBuffer.cpp (hunk 7 above)
Comment (tribal-tec) [reply engines/ospray/OSPRayFrameBuffer.cpp]: The pixelop code probably might need an adaption. I have to check that, good point. Not sure if I can test it though with my current setup though.

**Comment by tribal-tec** [comment]
I also agree that a server-side upscaling is definetly possible, but actually compressing and sending a smaller image (which in fact it is and it always was even before) makes more sense, plus as Juan mentioned, GPU texture scaling has no performance impact. The subsampling is an optional feature anyways, so clients can easily opt-out if they don't want to deal with that.

File: apps/ui/BaseWindow.cpp (hunk 3 above)
Comment (tribal-tec) [reply apps/ui/BaseWindow.cpp]: MAG is not needed :)

Comment (hernando) [reply apps/ui/BaseWindow.cpp]: No, no, it's MIN, I'm sure. You are mapping the texture to a bigger area.

**Comment by tribal-tec** [comment]
Updated

**Comment by tribal-tec** [comment]
retest this please

File: engines/ospray/OSPRayFrameBuffer.cpp (hunk 7 above)
Comment (tribal-tec) [reply engines/ospray/OSPRayFrameBuffer.cpp]: The 'normal' deflect streaming works, but the pixelop image is too small.

**Comment by hernando** [comment]
Daniel mentioned that Deflect streaming was not working but he didn't know what was the problem. I'd prefer if we merge this only when that's fixed. The question is whether we wait for him or somebody gives it a try.

**Comment by rolandjitsu** [comment]
We can wait 😄

**Comment by tribal-tec** [comment]
I added the DeflectPixelOp with an according fix for arbitrary (non TILE_SIZE) resolutions, which was the breaking part of this PR. This was an issue as the subsampling buffer can have any resolution in fact.

**Comment by tribal-tec** [comment]
So feel free to review (again) and finally approve :smile:

File: engines/ospray/OSPRayEngine.cpp [hunk 8]
```diff
@@ -198,11 +198,27 @@ void OSPRayEngine::preRender()
...
+        }
+        return result;
+    }
+
+    Vector2f result = size;
+    if (isSideBySideStereo)
+    {
+        if (size.x() % (TILE_SIZE * 2) != 0)
+            result.x() = size.x() - size.x() % (TILE_SIZE * 2);
+    }
```
Comment (hernando) [inline engines/ospray/OSPRayEngine.cpp]: Use the simpler code from the previous commit

**Pull request merged**